---
title: "Pre Process Spatial data"
date: "`r format(Sys.time(), '%d %B, %Y')`"
editor_options: 
  chunk_output_type: console

knit: (function(inputFile, out_dir, ...) {
    source("../bin/custom_knit_functions.R");
    knit_github(inputFile, "../lab_book/01_pre_processing/")
    })
---

# Spatial transcriptomics
***

```{r setup, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  fig.path="./Figures/",
  fig.align="center",
  fig.process = function(filename){
    new_filename <- stringr::str_remove(string = filename, 
                                        pattern = "-1")
    fs::file_move(path = filename, new_path = new_filename)
    ifelse(fs::file_exists(new_filename), new_filename, filename)
})

# output: 
#   github_document:
#     toc: yes

# setwd("/Users/vilkal/work/Brolidens_work/Projects/Spatial_DMPA/src/")
#.libPaths("/Users/vilkal/Applications/miniconda3/envs/scRNAseq2022/lib/R/library")
# reticulate::use_python("/Users/vilkal/Applications/miniconda3/envs/scRNAseq2022/bin/python", required = TRUE)
# reticulate::py_config()
```

### Load packages

```{r Load data, message=FALSE, warning=FALSE, include=FALSE}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(tidyseurat)
library(Seurat)
library(hdf5r)
#library(patchwork)
#library(Matrix)
# remotes::install_github("czarnewski/niceRplots",force=T)
library(niceRplots)

#library(SeuratData)
# devtools::install_github('satijalab/seurat-data')


```

### Load ST data

```{r}
#############
# LODA DATA #
#############
metadata <- read_csv("../data/Clinincal_data_Spatial_DMPA.csv")
sample_names <- metadata$ID
h5_files <- list.dirs(path = "../data",
                      full.names = T, recursive = T) %>%
            grep("outs_P10.$", ., value = TRUE) %>%
            set_names(., sample_names)

# Read in h5 files and create Seurat Object
seuratObj_list <- h5_files %>%
  map(.,  ~Load10X_Spatial(.x)) 

#################
# COLOR PALETTS #
#################

```

```{r, Clean up Seurat object}
####################################
# RENAME SAMPLES, SPOTS AND IMAGES #
####################################
seuratObj_list <- seuratObj_list %>%
  imap(., ~AddMetaData(object = .x, 
                       metadata = rep(.y, length(Idents(.x))), 
                       col.name = "orig.ident")) %>%
  map(.,  ~SetIdent(., value = .@meta.data$orig.ident)) %>%
  imap(., ~RenameCells(.x, 
                      new.names = paste0(.y,"_", gsub("-.*","",colnames(.x[["Spatial"]])))) ) %>%
  imap(., ~ {.x@images <- set_names(.@images,.y); .x})

##################
# MERGE SAMPLES #
#################
# Merge datasets into one single seurat object
DMPA  <- merge(seuratObj_list[[1]], y = seuratObj_list[2:length(seuratObj_list)])
```

# Create one merged object

We can now load the expression matricies into objects and then merge them into a single merged object. Each analysis workflow (Seurat, Scater, Scranpy, etc) has its own way of storing data. We will add dataset labels as cell.ids just in case you have overlapping barcodes between the datasets. After that we add a column `Chemistry` in the metadata for plotting later on.

Once you have created the merged Seurat object, the count matrices and individual count matrices and objects are not needed anymore. It is a good idea to remove them and run garbage collect to free up some memory.


# Quality control
***

Similar to scRNAseq we use statistics on number of counts, number of features and percent mitochondria for quality control. 

Now the counts and feature counts are calculated on the Spatial assay, so they are named  "nCount_Spatial" and "nFeature_Spatial".

```{r, fig.height=10}
####################
# QUALITY CONTROLL #
####################
DMPA <- PercentageFeatureSet(DMPA, "^MT-", col.name = "percent_mito")
DMPA <- PercentageFeatureSet(DMPA, "^HB[^(P)]", col.name = "percent_hb")


VlnPlot(DMPA, features = c("nCount_Spatial", "nFeature_Spatial","percent_mito","percent_hb"), 
        pt.size = 0.1, ncol = 2) + NoLegend()
```

We can also plot the same data onto the tissue section.

```{r, fig.height=12}

SpatialFeaturePlot(DMPA, features = c("nCount_Spatial", "nFeature_Spatial","percent_mito","percent_hb")) 

# nFeature is the number of uniqe genes in each spot
# nCount is the number of total counts in each spot 
```


As you can see, the spots with low number of counts/features and high mitochondrial content is mainly towards the edges of the tissue. It is quite likely that these regions are damaged tissue. You may also see regions within a tissue with low quality if you have tears or folds in your section. 

But remember, for some tissue types, the amount of genes expressed and proportion mitochondria may also be a biological features, so bear in mind what tissue you are working on and what these features mean.

### Filter

Select all spots with less than 25% mitocondrial reads, less than 20% hb-reads and 1000 detected genes. You must judge for yourself based on your knowledge of the tissue what are appropriate filtering criteria for your dataset.


```{r, Filter}
#############
# FILTERING #
#############
dim(DMPA)
DMPA %>%
  group_by(orig.ident) %>%
  summarise(across(c(nCount_Spatial:percent_hb), list(min = min, max = max))) %>%
  rename_with(., ~str_replace(., "_Spatial|percent_", ""))

# remove spots with; less than 500 genes, more than 25% mitochndrial genes, more than 20% hemoglobine genes:
DMPA <- DMPA[, DMPA$nFeature_Spatial>500 & DMPA$percent_mito < 25 & DMPA$percent_hb < 20]

dim(DMPA)
DMPA %>%
  group_by(orig.ident) %>%
  summarise(across(c(nCount_Spatial:percent_hb), list(min = min, max = max))) %>%
  rename_with(., ~str_replace(., "_Spatial|percent_", ""))

```

And replot onto tissue section:

```{r, fig.height=10}
SpatialFeaturePlot(DMPA, features = c("nCount_Spatial", "nFeature_Spatial","percent_mito"))

```

### Paulo section
```{r eval=FALSE, include=FALSE}
sample_names
niceRplots::plot_spatial_feat(DMPA, red = "P101", feat = "nCount_Spatial", plot_tissue = F, label = T)

niceRplots::plot_spatial_feat(DMPA, red = "P101", feat = c("nCount_Spatial", "percent_mito"), plot_tissue = F,)

#niceRplots::plot_spatial_feat(brain, red = "posterior1", feat = "nCount_Spatial", plot_tissue = F)
#niceRplots::plot_spatial_feat(brain, red = "anterior1", feat = "nCount_Spatial", plot_tissue = T, transparency = "00")
```




### Top expressed genes
As for scRNAseq data, we will look at what the top expressed genes are.

```{r eval=FALSE, fig.height=8, fig.width=6, include=FALSE}
C = DMPA@assays$Spatial@counts
C@x = C@x/rep.int(colSums(C), diff(C@p))
most_expressed <- order(Matrix::rowSums(C), decreasing = T)[20:1]
boxplot(as.matrix(t(C[most_expressed, ])), cex = 0.1, las = 1, xlab = "% total count per spot", 
    col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
```

As you can see, the mitochondrial genes are among the top expressed. Also the lncRNA gene Bc1 (brain cytoplasmic RNA 1). Also one hemoglobin gene. MALAT1 is a long-non-coding RNA

### Filter genes
We will remove the Bc1 gene, hemoglobin genes (blood contamination) and the mitochondrial genes.

```{r}
dim(DMPA)

# Filter MALAT1
DMPA <- DMPA[!grepl("MALAT1", rownames(DMPA)), ]

# Filter Mitocondrial
DMPA <- DMPA[!grepl("^MT-", rownames(DMPA)), ]

# Filter Hemoglobin gene (optional if that is a problem on your data)
DMPA <- DMPA[!grepl("^HB[^(P)]", rownames(DMPA)), ]

dim(DMPA)
```

### Top expressed genes after filtering

```{r eval=FALSE, fig.height=8, fig.width=6, include=FALSE}
C = DMPA@assays$Spatial@counts
C@x = C@x/rep.int(colSums(C), diff(C@p))
most_expressed <- order(Matrix::rowSums(C), decreasing = T)[20:1]
boxplot(as.matrix(t(C[most_expressed, ])), cex = 0.1, las = 1, xlab = "% total count per spot", 
    col = (scales::hue_pal())(20)[20:1], horizontal = TRUE)
```

# Analysis
***

For ST data, the Seurat team recommends to use SCTranform for normalization, so we will do that. `SCTransform` will select variable genes and normalize in one step.

```{r eval=FALSE, include=FALSE}

DMPA <- SCTransform(DMPA, assay = "Spatial", verbose = TRUE, method = 'poisson')

```

### Session info

```{r}
sessionInfo()
```
