---
title: "04_deconvolut_st_data"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    code-fold: show
params:
  fig.path: "./Figures/04/"
editor_options: 
  chunk_output_type: console
---

# Subset celltypes

```{r setup, include=FALSE}
knitr::opts_chunk$set(#fig.width = 6.6929133858,
                      fig.path="./Figures/",
                      fig.align="center",
                      message    = FALSE,
                      warning    = FALSE)
# setwd("/Users/vilkal/work/Brolidens_work/Projects/Spatial_DMPA/src")
```

```{r background_job, eval=FALSE, include=FALSE}
source("../bin/render_with_jobs.R")
file_name <- "./04_deconvolute_st_data.md"

file <- paste0(basename(xfun::sans_ext(file_name)), '_', Sys.Date(), '.html')

# quarto
# render_html_with_job(out_dir = lab_dir)
# fs::file_move(path = file, new_path = paste0(lab_dir, file))

# currently using quarto for github and kniter for html due to source code option 
render_git_with_job(fig_path = "../Figures/04/")

# kniter
knit_html_with_job(out_dir = "../lab_book/04_deconvolute_st_data", fig_path = "./Figures/04/")
```

## Load data and libraries

```{r Load_data}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(openxlsx)

library(RColorBrewer)
library(cowplot)
library(patchwork)

library(snowfall)
library(scran)
library(NMF)
library(gplots)
library(BiocParallel)

# library("devtools");
# remotes::install_github("czarnewski/niceRplots")
# remotes::install_github("renozao/xbioc") #, dependencies = FALSE
# remotes::install_github("meichendong/SCDC")
library(SCDC)
library(Biobase)
source("../bin/plotting_functions.R")

#########
# PATHS #
#########
input_dir_s <- "../results/03_clustering_st_data/"
input_dir_r <- "../results/04_DGE_ref_data/"
result_dir <- "../results/04_deconvolution_st_data/"
marker_dir <- "./marker_genes/"
bulk_path <- "/Users/vilkal/Raw_data/Bulk_Transcriptomics/Raw_counts_matrix.csv"
if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
if( isFALSE(dir.exists(marker_dir)) ) { dir.create(marker_dir,recursive = TRUE) }

#############
# LODA DATA #
#############
DATA_st <- readRDS(paste0(input_dir_s,"seuratObj_clustered.RDS"))
DATA_r <- readRDS(paste0(input_dir_r,"seuratObj_clustered.RDS"))
cell_id <- read.xlsx("../../Spatial_DMPA/data/Reference_data/annotation_single_cell.xlsx")
markers_genes <- readRDS(paste0(input_dir_r,"markers_genes.RDS"))

Bulk_data <-read_csv(bulk_path)

```

### SCDC deconcolution steps

1.  subset the reference dataset to get more equal number of cells for each cluster
2.  Normalize the data and run UMAP
3.  select top marker genes
4.  Create expression set
5.  run deconvolution

## Subset scRNA reference dataset

```{r subset_and_filter_ref_data}
# filter spots with too few reads
DATA_st_filt <- DATA_st  %>%
  filter(., nFeature_RNA > 200)

dim(DATA_st@assays$RNA@counts)
dim(DATA_st_filt@assays$RNA@counts)
#sum(DATA_st_filt@assays$RNA@counts[,"P031_TGCCTTGCCCTTACGG"])

# creating a subset ed object with 200 spots selected from each cluster
DATA_sub <- DATA_r %>%
    filter(., .cell %in% WhichCells(., downsample = 200)) %>%
    SetIdent(., value = "Clusters")
      
table(DATA_r$Clusters)
table(DATA_sub$Clusters)

# run normalization and dimensionality reduction
# only necessary to get new subset ed UMAP
DATA_sub <- DATA_sub %>%
  SCTransform(., ncells = 3000, verbose = FALSE, method = "poisson") %>% 
  RunPCA(verbose = FALSE) %>%
  RunUMAP(dims = 1:30)

p <- plot_clusters.fun(DATA_sub, cluster="Clusters", red = "UMAP")
```

## Select genes for deconvolution

```{r Get_best_markers}

# Identify the top genes that have a high difference in expression
top30 <- markers_genes %>%
  filter(.$gene %in% rownames(DATA_st@assays$RNA@counts)) %>%
  group_by(cluster) %>%
  top_n(-50, p_val_adj) %>%
  top_n(40, pct.diff) %>%
  top_n(30, log.pct.diff) 

# remove all VDJ-genes from list of HVG
# remove <- str_subset(top30$gene, "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG")
# top25_gr <- top25 %>%
#   ungroup() %>%
#   filter(., !(.$gene %in% remove)) %>%
#   group_by(cluster) %>%
#   group_split() %>%
#   set_names(., seq(0, length(.)-1) )

m_feats <- unique(as.character(top30$gene))
```

## Create ExpressionSet

```{r ExpressionSet}
# Create Expression Sets
eset_SC <- ExpressionSet(assayData = as.matrix(DATA_r@assays$RNA@counts[m_feats,
    ]), phenoData = AnnotatedDataFrame(DATA_r@meta.data))
eset_ST <- ExpressionSet(assayData = as.matrix(DATA_st_filt@assays$RNA@counts[m_feats,
    ]), phenoData = AnnotatedDataFrame(DATA_st_filt@meta.data))

bulk <- Bulk_data %>%
  select(-entrez, -length) %>%
  group_by(symbol) %>% 
  summarise_all(sum) %>%
  filter(!(is.na(.$symbol))) %>%
  filter(.$symbol %in% m_feats) %>%
  column_to_rownames(., var = "symbol")

eset_B <- ExpressionSet(assayData = as.matrix(bulk[m_feats,]))
```

## deconvolution

```{r deconvolution}
################
# BULK DECONV. #
#################
deconvolution <- SCDC::SCDC_prop(bulk.eset = eset_B, sc.eset = eset_SC, 
                                 ct.varname = "Clusters",
                                 ct.sub = as.character(unique(eset_SC$Clusters)))

saveRDS(deconvolution, paste0(result_dir,"deconvolution_bulk_scdc.RDS"))
# deconvolution <- readRDS(paste0(result_dir,"deconvolution_bulk.RDS"))

###################
# SPATIAL DECONV. #
###################
deconvolution <- SCDC::SCDC_prop(bulk.eset = eset_ST, sc.eset = eset_SC, 
                                 ct.varname = "Clusters",
                                 ct.sub = as.character(unique(eset_SC$Clusters)))

saveRDS(deconvolution, paste0(result_dir,"deconvolution_scdc.RDS"))
# deconvolution <- readRDS(paste0(result_dir,"deconvolution.RDS"))

head(deconvolution$prop.est.mvw)

# Create empty matrix and populate with SCDC results
temp <- Matrix::Matrix(0, 
                 nrow = ncol(deconvolution$prop.est.mvw), 
                 ncol = ncol(DATA_st), 
                 dimnames = list(colnames(deconvolution$prop.est.mvw), colnames(DATA_st)))

temp[,rownames(deconvolution$prop.est.mvw)] <- t(deconvolution$prop.est.mvw)

decon_column <- deconvolution$prop.est.mvw %>%
  DATA_st@assays$SCDC@data %>%
  as_tibble(., rownames = ".cell") %>%
  pivot_longer(., cols = -.cell, names_to = "decon_clus", values_to = "values") %>%
  filter(!(values==0)) %>% nest( data = -decon_clus)

DATA_st <- DATA_st %>% 
  left_join(., decon_column, by=".cell")

# take the deconvolution output and add it to the Seurat object as a new assay.
DATA_st@assays[["SCDC"]] <- CreateAssayObject(data = temp)
DATA_st@assays$SCDC@key <- "scdc_"

saveRDS(DATA_st, paste0(result_dir,"seuratObj_deconvolution_scdc.RDS"))
# DATA_st <- readRDS(paste0(result_dir,"seuratObj_deconvolution.RDS"))
# DATA_st <- readRDS(paste0(result_dir,"seuratObj_deconvolution_filt.RDS"))

```

```{r Add_cell_annotation}
cell_type <- c("fibroblasts","fibroblasts","immune","immune","immune","immune","immune","immune","immune","fibroblasts","immune","endothelial","epithelial","immune","fibroblasts","fibroblasts","immune","immune","immune","fibroblasts","immune","epithelial","fibroblasts","granulocytes","immune","immune","endothelial","immune","immune") %>% set_names(., as.character(0:28))

annot2 <- c("fibroblasts","fibroblasts","T-cell","T-cell","T-cell","T-cell","T-cell","T-cell","T-cell","fibroblasts","T-cell",NA,NA,"T-cell","fibroblasts","fibroblasts","myeloid",NA,"myeloid",NA,NA,NA,NA,NA,"T-cell","B cell",NA,"T-cell", NA) %>% set_names(., as.character(0:28))

tibble(cell_type)

DATA_st %>%
  filter()

df <- DATA_st@assays$SCDC@data %>%
  as_tibble(., rownames = "clus") %>%
  pivot_longer(., cols = -clus, names_to = ".cell", values_to = "values") %>%
  filter(!(values==0)) %>% 
  #nest( data = -clus) %>%
  left_join(., select(as_tibble(DATA_st),.cell, Clusters, groups, sp_annot), by=".cell") %>%
  mutate(., cell_type = cell_type[as.character(.$clus)]) %>%
  mutate(., cell_type_2 = annot2[as.character(.$clus)]) %>%
  mutate(clus = factor(clus, levels = as.character(0:28))) 

df_ <- df %>%
    #mutate(orig.ident = !!(facet)) %>%
    #mutate(orig.ident = factor(.data[["orig.ident"]], levels = lvls)) %>%
    #cbind(.,as_tibble(select(spe, groups))) %>%
  mutate(cell_type_2 = case_when(.$cell_type == "epithelial" & .$Clusters == "9" ~ 'ker_1',
                                 .$cell_type == "epithelial" & .$Clusters == "8" ~ 'ker_1',
                                 .$cell_type == "epithelial" & .$Clusters == "7" ~ 'ker_1',
                                 .$cell_type == "epithelial" & .$Clusters == "6" ~ 'ker_1',
                                 FALSE ~ .$cell_type_2))
  #mutate(cell_type_2 = ifelse(grepl("9|8|7|6", .$Clusters), ) 
    
 
kreatins <- decon_column %>% 
  filter(cell_type == "epithelial") 
  #mutate(Keratinocytes = ifelse(grepl("9|8|7", .$Clusters), ) 
```

```{r waist_of_time, include=FALSE, eval=FALSE}
DATA_st <- DATA_st %>% mutate(sp_annot2 = ifelse(.$sp_annot == "epi_1"|.$sp_annot == "epi_2"|.$sp_annot == "epi_3", "epi", .$sp_annot )) %>%
    mutate(sp_annot2 = ifelse(.$sp_annot2 == "SubMuc_1"|.$sp_annot2 == "SubMuc_2"|.$sp_annot2 == "SubMuc_3", "SubMuc", .$sp_annot2 ))

find_hull <- function(df) df[chull(df$eff, df$man), ]

df_ <- df %>%
  cbind(.,as_tibble(select(DATA_st, Clusters))) %>%
  mutate(Clusters = as.character(.$Clusters)) %>%
  cbind(.,as_tibble(select(spe, orig.ident))) %>%
    #mutate(orig.ident = !!(facet)) %>%
    #mutate(orig.ident = factor(.data[["orig.ident"]], levels = lvls)) %>%
    #cbind(.,as_tibble(select(spe, groups))) %>%
    rownames_to_column(var = "barcode") %>%
    as_tibble() %>%
    filter(orig.ident == "P118")

get_chull <- function(df_){
  chull <- df_[grDevices::chull(df_$imagecol, df_$imagerow), ][["barcode"]]
  chull <- ifelse(df_$barcode %in% chull, "c", NA)
  return(chull)
}
      
  df_ <- df_ %>% 
  nest(data= -Clusters) %>%
  mutate(chull = map(.$data, ~get_chull(.x))) %>%
  unnest(c(data, chull))


hull <- df_ %>%
  filter(!(is.na(.$chull)))

p <- ggplot(data=df_, aes(x=imagecol,y=imagerow)) +
      ggstar::geom_star(data=df_, aes(x=imagecol,y=imagerow, fill=Clusters, colour=Clusters),
        starshape = "hexagon",
        size = 2.8,
        #stroke = 0,
        alpha = alpha
      )+ spatial_annotation + 
      geom_point(aes(x=imagecol,y=imagerow, colour=Clusters),
               stroke = 0, size = point_size, alpha = alpha) +
      geom_polygon(data=hull, alpha =0.3, aes(group = Clusters, fill=Clusters))
  
      geom_voronoi_tile(aes(fill = Clusters), colour = NA, max.radius = 0.4)
      geom_voronoi_tile(aes(fill = Clusters), colour = NA) +
      geom_voronoi_segment() +
      
      
      #scale_colour_identity() +
      scale_color_gradientn(colours = col, 
                            values = seq(from=0, to=1, along.with=col),
                            na.value = "#FFFFFF") + 
      spatial_image + 
      spatial_annotation + 
      #scale_color_manual(values=c(annot_col, "transparent")) +
      coord_cartesian(expand=FALSE ) + #theme(l) +
      xlim(l$min_col,l$max_col) +
      ylim(l$max_row,l$min_row) +
      geom_text(aes(label = sample_id, x=x, y=y), data = text_annot, inherit.aes = F) + # sample ID
      facet_wrap(~factor(orig.ident, levels = lvls), ncol = ncol)
      
ggplot(iris, aes(Sepal.Length, Sepal.Width, group = -1L)) +
  geom_point(aes(colour=Species))
  geom_voronoi_tile(aes(fill = Species)) +
  geom_voronoi_segment() +
  geom_text(aes(label = after_stat(nsides), size = after_stat(vorarea)),
    stat = 'delvor_summary', switch.centroid = TRUE
  )
```

```{r cells_per_cluster}
################
# ST BAR PLOT #
################
decon_column <- DATA_st@assays$SCDC@data %>%
  as_tibble(., rownames = "clus") %>%
  pivot_longer(., cols = -clus, names_to = ".cell", values_to = "values") %>%
  filter(!(values==0)) %>% 
  #nest( data = -clus) %>%
  left_join(., select(as_tibble(DATA_st),.cell, Clusters, groups, sp_annot), by=".cell") %>%
  mutate(sp_annot2 = ifelse(.$sp_annot == "epi_1"|.$sp_annot == "epi_2"|.$sp_annot == "epi_3", "epi", .$sp_annot )) %>%
  mutate(sp_annot2 = ifelse(.$sp_annot2 == "SubMuc_1"|.$sp_annot2 == "SubMuc_2"|.$sp_annot2 == "SubMuc_3", "SubMuc", .$sp_annot2 )) %>%
  filter(!(is.na(.$sp_annot2))) %>% 
  mutate(., cell_type = cell_type[as.character(.$clus)]) %>%
  mutate(., cell_type_2 = annot2[as.character(.$clus)]) %>%
  mutate(clus = factor(clus, levels = as.character(0:28)))
  

n <- decon_column %>%
  select(clus, Clusters, cell_type, cell_type_2) %>%
  #unique() %>%
  group_by(Clusters) %>%

  nest() %>%
  ungroup() %>%
  #tally()
  map(.$data, ~count(.x))
  mutate(n = map(.$data, ~enframe(table(.x))))
  
  
kreatins <- decon_column %>% 
  filter(cell_type == "epithelial") 
  #mutate(Keratinocytes = ifelse(grepl("9|8|7", .$Clusters), )
         
percent.fun <- function(obj, var, name){
  for(i in length(var)){
    obj <- PercentageFeatureSet(obj, var[i], col.name = name[i])
  }
  return(obj)
}

genes <- c("KRT15", "KRT14", "KRT10", "KRT5", "KRT1", "LOR", "CDH1", "CD4") 
genes <- c( "CD4") 
ker <- percent.fun(DATA_st, ker, ker) %>%
  select(Clusters, nCount_RNA, any_of(c("KRT15", "KRT14", "KRT10", "KRT5", "KRT1"))) %>%
  sum(.[,"rna_KRT1"])

ker <- DATA_st %>% 
  #mutate(., FetchData(., vars = genes, slot = "counts" )) %>%
  mutate(., FetchData(., vars = genes, slot = "data" )) %>%
  select(groups, Clusters, nCount_RNA, any_of(genes)) %>% # nCount_RNA,
  group_by(Clusters, groups) %>%
  summarise(across(where(is_numeric), ~sum(.))) %>%
  ungroup() %>%
  #mutate(across(any_of(genes), ~./.data[["nCount_RNA"]])) %>%
  select(-nCount_RNA) %>%
  pivot_longer(cols = -any_of(c("Clusters", "groups")), names_to="Genes", values_to = "values")

ggplot(ker, aes(fill=Genes, y=values, x=Clusters)) + 
    geom_bar(position="stack", stat="identity") +
    #scale_y_continuous(labels=scales::percent) +
    facet_wrap(~groups) + theme_minimal()

ggplot(decon_column, aes(fill=clus, y=values, x=Clusters)) + 
    geom_bar(position="stack", stat="identity") +
  scale_fill_manual(values = pal) +
  facet_wrap(~groups + sp_annot2) + theme_minimal() 
```

```{r Colour_pallet_sub}
cell_type <- c("fibroblasts","fibroblasts","immune","immune","immune","immune","immune","immune","immune","fibroblasts","immune","endothelial","epithelial","immune","fibroblasts","fibroblasts","immune","immune","immune","fibroblasts","immune","epithelial","fibroblasts","granulocytes","immune","immune","endothelial","immune","immune") %>% set_names(., 0:28)

 df <- tibble(cell_id =rep(0:28, 30)) %>%
   mutate(cell_name = cell_type[.$cell_id])


ColourPalleteMulti <- function(df, group, subgroup){

  # Find how many colour categories to create and the number of colours in each
  categories <- aggregate(as.formula(paste(subgroup, group, sep="~" )), df, function(x) length(unique(x)))
  s <- c(RColorBrewer::brewer.pal(8,"Set2"), RColorBrewer::brewer.pal(8,"Set1"))
  e <- c(RColorBrewer::brewer.pal(8,"Pastel2"), RColorBrewer::brewer.pal(8,"Pastel1"))
  category.start <- (s[1:nrow(categories)]) # Set the top of the colour pallete
  category.end  <- (e[1:nrow(categories)]) # set the bottom

  # Build Colour pallette
  colours <- unlist(lapply(1:nrow(categories),
                          function(i){
                            colorRampPalette(colors = c(category.start[i],
                                                        category.end[i]))(categories[i,2])}))
  return(colours)
}
# choose higer and lower level:
# colours <- ColourPalleteMulti(candida_df, "rank", "taxName") %>% set_names(levels(candida_df$taxName))


scales::show_col(scales::hue_pal()(8))
```

```{r bulk_cells}
#################
# BULK BAR PLOT #
#################
get_order <- function(df, value){
  value <- enquo(value) 
  df %>%
    group_by(ID) %>%
    add_tally(wt = !!(value), sort = F) %>%
    ungroup() %>%
    mutate(ID = fct_reorder(ID, n))
    
}
get_order2 <- function(df, value_1, value_2){
  value_1 <- enquo(value_1)
  value_2 <- enquo(value_2)
  df %>%
    group_by(ID) %>%
    add_tally(wt = !!(value_1), sort = F) %>%
    add_tally(wt = !!(value_2), sort = F, name = "n2") %>%
    ungroup() %>%
    mutate(ID = fct_reorder2(ID, n, n2))
    
}
get_legend.fun <- function(df, title, colours){
  # https://stackoverflow.com/questions/73711918/how-to-remove-majority-of-blank-space-from-get-legend
  p <- ggplot() + geom_col(data = df, aes(fill=annot, y=values, x=ID)) +
  scale_fill_manual(title, values = colours[unique(pull(df, "annot"))]) 
  l <- cowplot::get_legend(p)
  return(l)
}

grid::grid.draw(c$legend[[1]])


id <- c("P107", "P108", "P114", "P097","P118", "P105", "P080", "P031")  
gr <- c("DMPA", "DMPA", "DMPA", "DMPA", "ctrl", "ctrl", "ctrl", "ctrl") %>% set_names(., id)
gr <- set_names(meta$Contraception, meta$ID) 

decon_column <- deconvolution$prop.est.mvw %>%
  as_tibble(., rownames = "ID") %>%
  pivot_longer(., cols = -ID, names_to = "decon_clus", values_to = "values") %>%
  mutate(decon_clus = factor(decon_clus, levels = as.character(0:28))) %>%
  #filter(ID %in% DATA_st$orig.ident) %>%
  mutate(., group = gr[as.character(.$ID)])  %>%
  mutate(., cell_type = cell_type[as.character(.$decon_clus)]) %>%
  mutate(., cell_type_2 = annot2[as.character(.$decon_clus)]) %>%
  mutate(., annot = paste0(.$decon_clus, "_", cell_type_2)) %>%
  mutate(., cell_type_2 = case_when(is.na(.$cell_type_2) ~ .$cell_type, TRUE ~ .$cell_type_2)) %>%
  mutate(., annot = case_when(is.na(.$cell_type_2) ~ paste0(.$decon_clus, "_", .$cell_type), TRUE ~ .$annot))
  
# arrange order 
df <- decon_column %>%
  arrange(cell_type, cell_type_2) %>%
  mutate(annot = factor(.$annot, levels = unique(.$annot))) %>%
  mutate(decon_clus = factor(.$decon_clus, levels = as.character(seq(1:28)))) %>%
  mutate("14" = ifelse(.$decon_clus == "14", .$values, NA)) %>%
  mutate("18" = ifelse(.$decon_clus == "18", .$values, NA)) %>%
  #get_order2(., `18`, `14`)
  get_order(., `14`)

colours <- ColourPalleteMulti(df, "cell_type", "annot") %>% set_names(levels(decon_column$annot))
# scales::show_col(colours)
c <- df %>%
  nest(data = -cell_type) %>%
  mutate(legend = pmap(., ~get_legend.fun(..2, ..1, colours))) %>%
  

# colours <- c(scales::hue_pal()(8),
#              RColorBrewer::brewer.pal(9,"Set1"),
#              RColorBrewer::brewer.pal(8,"Set2"),
#              RColorBrewer::brewer.pal(8,"Accent"),
#              RColorBrewer::brewer.pal(9,"Pastel1"),
#              RColorBrewer::brewer.pal(8,"Pastel2") )

bar <- ggplot(df, aes(fill=annot, y=values, x=ID)) + 
    geom_bar(position="stack", stat="identity") +
    theme_minimal() +
    scale_fill_manual(values = colours) +
    theme(axis.text.x = element_text(angle = 45, vjust = 0.5, hjust=1),
          legend.position = "none") +
    #facet_wrap(~group ) +
    facet_grid(cols = vars(group), scales = "free_x", space = "free_x", switch = "x")

l <- plot_grid(plotlist = c$legend, nrow =3 , byrow = T , rel_heights = c(1,1,3), rel_widths = c(.5,0.2), align = "hv", axis = "t")
plot_grid(bar, l ,rel_widths = c(1,.2))
  #axis = "tblr",
#################
# BULK HEATMAP #
#################

```

## Heatmap of bulk deconvolution

```{r plot Heatmap, fig.width = 3, fig.height = 3, fig.dim = c(33, 33), eval=FALSE}
#install_github("jokergoo/ComplexHeatmap")
# library(ComplexHeatmap)
#################
# COLOUR PALLET #
#################
col.pal <- function(df, col, n) {
          lable <- seq(max(df, na.rm = T), min(df, na.rm = T), length = n)
          #lable <- append(lable, 0, after = pos)
          col <- colorRampPalette(c(col))(n)
          #col <- brewer.pal(n=n, name=pal)
          #col <- append(col, "white", after = pos)
          col_fun = colorRamp2(lable, col)
}

# Colour pallet:
# lable <- seq(max(zscore), min(zscore), length = 8)
# lable <- append(lable, 0, after = 4)
# col <- brewer.pal(n=8, name="RdYlBu")
# col <- append(col, "white", after = 4)
# col_fun = colorRamp2(lable, col)

ptc_ss <- c("#364B9A", "#4A7BB7", "#6EA6CD", "#98CAE1", "#C2E4EF", "#EAECCC", "#FEDA8B", "#FDB366", "#F67E4B", "#DD3D2D", "#A50026")

##########
# MATRIX #
##########
matrix <- deconvolution$prop.est.mvw %>%
  as_tibble(., rownames = "ID") %>%
  arrange(ID) %>%
  pivot_longer(., cols = -ID, names_to = "decon_clus", values_to = "values") %>%
  group_by(decon_clus) %>%
  mutate(zscore = (values - mean(values))/sd(values)) %>%
  select(-values) %>%
  pivot_wider(., -c(ID), names_from = ID, 
              values_from = zscore, values_fn = sum ) %>%
  column_to_rownames(var = "decon_clus") %>%
  as.matrix()

# z-scores
zscore <- t(apply(matrix,1,function(i){scale(i,T,T)}))
colnames(zscore) <- colnames(matrix)
zscore[is.nan(zscore)] <- 0

##############
# ANNOTATION #
##############
annot <- meta %>%
  filter(Trx =="yes") %>%
  select(ID, HIVstatus, Contraception, Tissue_gr, BV_Diagnosis_v3) %>%
  column_to_rownames(var = "ID") %>%
  arrange(ID)

anno <- annot %>%
  HeatmapAnnotation(df = ., 
                    col = list(
                      Contraception = c("DMPA"='#FB5273', "none"='#4FCEEF'),
                      Tissue_gr = c("T1"="#0072B2", "T2"="#009E73", "T3"="#D55E00", "T4"="#CC79A7", "T5"="#E69F00"),
                      BV_Diagnosis_v3 = c("BV"="tomato", "Interm"="orange","Normal"="#d7d7d7"),
                      HIVstatus = c("neg"="#d7d7d7","pos"="tomato") 
                             ))
#colorRamp2(seq(0, 2, length.out=length(ptc_ss)), ptc_ss) strcture(ptc_ss) ptc_rnbwdcolorRamp2(c(0, 1, 2), c("blue", "white", "red"))
dim(matrix)
dim(annot)
#################
# PLOT HEATMAP #
################
(H <- Heatmap(matrix, 
             #col = genes.col,
             top_annotation = anno,
             show_column_names = T,
             show_row_names = T,
             #clustering_distance_rows= "spearman",
             #clustering_distance_columns=  "spearman",
             column_names_gp = grid::gpar(fontsize = 4),
             column_names_rot = 47,
             #direction = "horizontal"
             heatmap_legend_param = list(
               title = "Cell proportion \n(scaled)"
               #at = c(-4,-2,0,2, 4), 
               #labels = c("0.4%", "0.3%", "0.2%", "0.1%", "0%")
               )
             ) )
draw(H, merge_legend = TRUE)
```

## Paulos code

```{r Paulos_facet_wrap_plotting, eval=FALSE, include=FALSE}
start_time <- Sys.time()

img_list <- names(DATA_st@images)
clusters <- rownames(DATA_st@assays$SCDC@data)


if(!dir.exists(result_dir)){dir.create(result_dir,recursive = T)}

library(niceRplots)

pdf(paste0(result_dir, "_HE.pdf"),width = (length(img_list) + 1)*3.5,height = 3.5,useDingbats = F)
par(mfrow= c(1,length(img_list) + 1))
plot_meta( DATA_sub , feat = "Clusters" , cex=.5,label = T, red = "umap_harmony" )
for(i in img_list){
  plot_spatial_feat(DATA_st, red = i,feat =  "CDH1", main="",assay="RNA", # CDH1, "ACTB"
                    pch=16,cex=1,plot_tissue = T,
                    transparency = "00")}
dev.off()

grid_genes <- function(plot_list, title){
  title <- ggdraw() + draw_label(paste0("Top 10 Markers for Cluster ", title), fontface='bold')
  g <- plot_grid(plotlist = plot_list,
            ncol = 4)
  g_ <- plot_grid(title, g, ncol=1, rel_heights=c(0.1, 1))
  return(g_)
}

for(j in clusters){
  pdf(paste0(result_dir,"cluster",j,".pdf"),
      width = (length(img_list) + 1)*3.5,
      height = 3.5,useDingbats = F)
  
  par(mfrow= c(1,length(img_list) + 1))
  DATA_sub$temp <- (DATA_sub$Clusters == j)*1
  plot_feat( DATA_sub, feat = "temp", red = "umap_harmony",  cex=.5, main=j )
  
  for(i in img_list){
    plot_spatial_feat(DATA_st, 
                      red = i, 
                      feat =  j, 
                      main=paste0(i," (",j,")"),
                      assay="SCDC",
                      maxs=max(DATA_st@assays$SCDC@data[j,],0.1),
                      pch=16,cex=1, 
                      plot_tissue = F)
  }
dev.off()
}
end_time <- Sys.time()
time <- end_time - start_time
```

```{r plot_clusters, dev.args = list(bg = 'transparent')}
######################
# FACET WRAP CLUSTER #
#####################
# I want to plot the intensity of the cell population in 
clus_exp <- function(spe, clus_spe, clusters, gene, assay="RNA"){
  #c <- unique( pull(spe, as_label(clusters)))
  col=c("grey90","grey80","grey60", "blue4","navy","black")
  clus_col <- c("grey90","navy")
  clusters <- enquo(clusters)
  c <- unique( pull(clus_spe, !!(clusters))) %>% sort(.) %>% set_names()
    
  # Create Columns for induvidual clusters:
  clus_columns <- map(c, ~(clus_spe@meta.data[["Clusters"]] == .x)*1) %>% 
    map(.,~as.character(.x)) %>%
    set_names(., paste0("Clus_",names(.)))
  
  # Add the Cluster columns to seurat obj.
  clus_spe <- bind_cols(clus_spe, clus_columns)
  # Plot Individal plot for each cluster
  P <- {map(c, ~plot_clusters.fun(clus_spe, color=clus_col, paste0("Clus_",.x), 
                                  red="umap_harmony"), txt_size=20, dot_size = 0.2, lable=F)}
  #clus_plot <- plot_clusters.fun(spe, color=col, paste0("Clus_","2"), red="umapharmony", txt_size=20, lable=F)
  
  filt_empty<- function(x, n_spot = 0) {x[c(TRUE, colSums(select(x, !(.cell))) > n_spot)]}
  
  decon_columns <- spe@assays$SCDC@data %>% 
    Matrix::t() %>%
    as_tibble(., rownames = ".cell") %>% 
    mutate_if(is.numeric, ~1 * (. > 0)) %>%
    filt_empty(., 0) %>%
    mutate(across(-.cell, ~as.character(.x)))
    
    c <- intersect(as.character(c), colnames(decon_columns))
    
  PLOTS <- tibble("clus" = c) %>%
    #sample_n(., size = 2) %>%
    mutate(plots = pmap(.,
      ~plot_st_feat.fun( spe,  # filter(spe, decon_columns[[..1]]=="1"), # removes spots with no % of given cell type
        assay=assay,
        geneid = ..1,#"KRT15", #"PTPRC",#"sp_annot",#"CDH1",
        zoom = "zoom", #"zoom"
        col = col,
        maxs = max(spe@assays$SCDC@data[..1,],0.1),
        #annot_col = "#dbd9d9",
        annot_line = .1,
        img_alpha = 0,
        point_size = 0.5)
      )) %>% #PLOTS <- PLOTS %>%
    mutate(l = map(.$plots, ~length(.x))) %>%
    #mutate(plots = )
    mutate(clus_plot = ifelse(.$clus %in% names(P), P[.$clus], NA)) %>%
    mutate(., combined = pmap(.,
      ~plot_grid( ..4, ..2, ncol=2, rel_widths=c(2,4)) ))

  # PLOTS$combined[[1]]
  #PLOTS$data[[1]]

  combined <- plot_grid( plotlist=PLOTS$combined, ncol=1, rel_heights=rep_len(1, length(PLOTS$combined) ) )

  return(combined)
}
Key(DATA_st@reductions$umap_harmony) <- "umapharmony_"

# spe <- DATA_st
# clus_spe <- DATA_r
# clusters <- sym("Clusters")
# gene <- "CDH1"
# assay <- "SCDC"
# clus_col <- c("grey90","navy")
#gr <- c("P107", "P108", "P114", "P097","P118", "P105", "P080", "P031")
start_time <- Sys.time()
CLUS_PLOTS <- clus_exp(DATA_st, DATA_r,clusters=Clusters, assay="SCDC")
end_time <- Sys.time()
time_n <- end_time - start_time
ggsave("./clus_plots.pdf",CLUS_PLOTS, height = 60, width = 10, limitsize = FALSE)

plot_grid( PLOTS$clus_plot[[1]], PLOTS$plots[[1]], ncol=2, rel_widths=c(2,4)) 
```

# BayesPrism

```{r BayesPrism, include=FALSE, eval=FALSE}
######################
# LOAD SPARSE MATRIX #
######################
# install_github("Danko-Lab/BayesPrism/BayesPrism")
library(BayesPrism)
data_dir <- "../data/Reference_data/GSE173231/"
f <- c("matrix.mtx$", "genes.tsv$|features.tsv$", "barcodes.tsv$") %>% set_names()

f_list <- map_df(f, ~list.files(path = data_dir, pattern = .x,
                      full.names = T, recursive = T) ) %>%
  mutate(sample_name = str_match(.$"matrix.mtx$", ".*\\/([^\\/]+).matrix\\.mtx$")[,2]) %>%
  mutate(sample_id = str_extract(.$sample_name, "(?<=_).+(?=_)"), .$sample_name) 

matrix_list <- pmap(f_list, ~ReadMtx(mtx = ..1, features = ..2, cells = ..3, 
                                     feature.column = 1) ) %>% # , mtx.transpose=T
                     set_names(., f_list$sample_id) %>% 
  map(., ~t(.x)) %>%
  imap(., ~`rownames<-`(., paste0(str_extract(rownames(.x), "^[ATCG]+"),"_",.y)))


#################################
# substitute symbols for entrez #
#################################
genes <- f_list$`genes.tsv$|features.tsv$` %>%
  set_names(., f_list$sample_id) %>% 
  map(., ~read_tsv(.x, col_names = F)) %>%
  bind_rows(., .id = "sample_id")
  
genes %>% 
  dplyr::group_by(X2) %>%
  dplyr::summarise(n = dplyr::n(), .groups = "drop") %>%
  dplyr::filter(n > 10) 

gene_list <- set_names(genes$X2, genes$X1)


DATA_r@assays$RNA@counts %>%
  t() %>%
  colnames()
  
t <- tibble(symbol = dimnames(DATA_r@assays$RNA@counts)[[1]]) %>%
  left_join(., select(genes, entrez="X1", symbol="X2"), by="symbol") %>%
  unique() %>%
  arrange(dimnames(DATA_r@assays$RNA@counts)[[1]])
  unique()
  mutate(., genes = cell_type[as.character(DATA_sub$Clusters)])



matrix_list %>%  
  bind_rows()
  
table(DATA_r@active.ident)
matrix_list$CX1@Dimnames[[1]]


###################################
# Get scRNA-seq  and filter genes #
###################################
sc.dat <- DATA_r %>%
  GetAssayData(., slot = 'counts', assay = "RNA") %>%
  # .[rownames(.) %in% Bulk_data$symbol, ] %>% # filtering is not neccesary done by BayesPrisem
  t(.) %>%
  as.matrix(.)

################################
# Get Cell type for scRNA-seq #
###############################
cell_type <- c("fibroblasts","fibroblasts","immune","immune","immune","immune","immune","immune","immune","fibroblasts","immune","endothelial","epithelial","immune","fibroblasts","fibroblasts","immune","immune","immune","fibroblasts","immune","epithelial","fibroblasts","granulocytes","immune","immune","endothelial","immune","immune") %>% set_names(., seq_along(.))

Clustering <- DATA_r@meta.data %>%
  mutate(., cell_type = cell_type[as.character(DATA_r$Clusters)]) %>%
  select(Clusters, cell_type) 

cell.state.labels <- Clustering$Clusters
cell.type.labels <- Clustering$cell_type

table(cell.state.labels)
table(cell.type.labels)

table(cbind.data.frame(cell.state.labels, cell.type.labels))
# Please make sure that all cell states contain a reasonable number of cells, e.g. >20 or >50, so that their profile can be represented accurately
##############
# BULK DATA #
#############
# entrez
bk.dat <- Bulk_data %>%
  select(symbol, contains("P")) %>%
  # mutate(entrez = paste0("ENSG",str_pad(as.character(Bulk_data$entrez), 11, pad = "0"))) %>%
  pivot_longer(-symbol) %>% 
  pivot_wider(names_from=symbol, values_fn = function(x) sum(x)) %>% #aggregates the duplicates
  column_to_rownames(var = "name")

####################################
# QC of cell type and state labels #
####################################
plot.cor.phi(input=sc.dat,
                         input.labels=cell.state.labels,
                         title="cell state correlation",
                         #specify pdf.prefix if need to output to pdf
                         #pdf.prefix="gbm.cor.cs", 
                         cexRow=0.2, cexCol=0.2,
                         margins=c(2,2))

plot.cor.phi(input=sc.dat, 
                         input.labels=cell.type.labels, 
                         title="cell type correlation",
                         #specify pdf.prefix if need to output to pdf
                         #pdf.prefix="gbm.cor.ct",
                         cexRow=0.5, cexCol=0.5,
                         )

sc.stat <- plot.scRNA.outlier(
  input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID 
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE #return the data used for plotting. 
  #pdf.prefix="gbm.sc.stat" specify pdf.prefix if need to output to pdf
)

bk.stat <- plot.bulk.outlier(
  bulk.input=bk.dat,#make sure the colnames are gene symbol or ENSMEBL ID 
    sc.input=sc.dat, #make sure the colnames are gene symbol or ENSMEBL ID 
  cell.type.labels=cell.type.labels,
  species="hs", #currently only human(hs) and mouse(mm) annotations are supported
  return.raw=TRUE
  #pdf.prefix="gbm.bk.stat" specify pdf.prefix if need to output to pdf
)

#################
# Filter genes #
################
sc.dat.filtered <- cleanup.genes (input=sc.dat,
                                  input.type="count.matrix",
                                    species="hs", 
                                    gene.group=c( "Rb","Mrp","other_Rb","chrM","MALAT1","chrX","chrY") ,
                                    exp.cells=5)
setdiff(colnames(sc.dat.filtered), colnames(bk.dat))

plot.bulk.vs.sc (sc.input = sc.dat.filtered,
                            bulk.input = bk.dat
                            #pdf.prefix="gbm.bk.vs.sc" specify pdf.prefix if need to output to pdf
)
# select only protein coding genes
sc.dat.filtered.pc <-  select.gene.type (sc.dat.filtered,
                                        gene.type = "protein_coding")

############################
# Construct a prism object #
############################
myPrism <- new.prism(
  reference=sc.dat.filtered.pc, 
  mixture=bk.dat,
  input.type="count.matrix", 
  cell.type.labels = cell.type.labels, 
  cell.state.labels = cell.state.labels,
  key=NULL,
  outlier.cut=0.01,
    outlier.fraction=0.1,
)
##################
# RUN BAYESPRISM #
##################
bp.res <- run.prism(prism = myPrism, n.cores=50)
```

```{r old_code, include=FALSE, eval=FALSE}
# This was an earlier version of the clus_plot function before I completely understod what to do
# might still be interesting but could also be just crap, Plots the the gene expression of a specific gene per cell type defined by deconvolution
###################
# Facet wrap GENE #
###################
clus_exp <- function(spe, clus_spe, clusters, gene, assay="RNA"){
  #c <- unique( pull(spe, as_label(clusters)))
  col=c("grey90","grey80","grey60","navy","black")
  clus_col <- c("grey90","navy")
  clusters <- enquo(clusters)
  c <- unique( pull(clus_spe, !!(clusters))) %>% sort(.) %>% set_names()
    
  # Create Columns for induvidual clusters:
  clus_columns <- map(c, ~(clus_spe@meta.data[["Clusters"]] == .x)*1) %>% 
    map(.,~as.character(.x)) %>%
    set_names(., paste0("Clus_",names(.)))
  
  # Add the Cluster columns to seurat obj.
  clus_spe <- bind_cols(clus_spe, clus_columns)
  # Plot Individal plot for each cluster
  P <- {map(c, ~plot_clusters.fun(clus_spe, color=clus_col, paste0("Clus_",.x), 
                                  red="umap_harmony"), txt_size=20, dot_size = 0.2, lable=F)}
  #clus_plot <- plot_clusters.fun(spe, color=col, paste0("Clus_","2"), red="umapharmony", txt_size=20, lable=F)
  
  filt_empty<- function(x, n_spot = 0) {x[c(TRUE, colSums(select(x, !(.cell))) > n_spot)]}
  
  decon_columns <- spe@assays$SCDC@data %>%
    Matrix::t() %>%
    as_tibble(., rownames = ".cell") %>% 
    mutate_if(is.numeric, ~1 * (. > 0)) %>%
    filt_empty(., 0) %>%
    mutate(across(-.cell, ~as.character(.x)))
    
    c <- intersect(as.character(c), colnames(decon_columns))
  
  PLOTS <- tibble("clus" = c) %>%
    #sample_n(., size = 2) %>%
    mutate(plots = pmap(.,
      ~plot_facets.fun(filter(spe, decon_columns[[..1]]=="1"),
        assay=assay,
        geneid = gene,#"KRT15", #"PTPRC",#"sp_annot",#"CDH1",
        zoom = "zoom", #"zoom"
        colors = col,
        #annot_col = "#dbd9d9",
        annot_line = .1,
        img_alpha = 0,
        point_size = 0.5)
      )) %>%
    mutate(clus_plot = ifelse(.$clus %in% names(P), P[.$clus], NA)) %>%
    mutate(., combined = pmap(.,
      ~plot_grid( ..3, ..2, ncol=2, rel_widths=c(1.2,4)) ))

  # PLOTS$combined[[1]]
  #PLOTS$data[[1]]

  combined <- plot_grid( plotlist=PLOTS$combined, ncol=1, rel_heights=rep_len(1, length(PLOTS$combined) ) )

  return(combined)
}
```

## Session info

```{r}
sessionInfo()
```
