---
title: "Differential gene expression"
date: "`r format(Sys.time(), '%d %B, %Y')`"
editor_options: 
  chunk_output_type: console
knit: (function(inputFile, out_dir, ...) {
    source("../../bin/custom_knit_functions.R");
    knit_github(inputFile, "../../lab_book/04_DGE_reference_data")
    })
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(#fig.width = 6.6929133858,
                      fig.path="./Figures/",
                      fig.align="center",
                      message    = FALSE,
                      warning    = FALSE)
# setwd("/Users/vilkal/work/Brolidens_work/Projects/Spatial_DMPA/src/00_reference_data")
```

## Load data and libraries
```{r}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(cowplot)

#########
# PATHS #
#########
input_dir <- "../../results/03_clustering_ref_data/"
result_dir <- "../../results/04_DGE_ref_data/"
marker_dir <- "./Marker_genes/"
if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
if( isFALSE(dir.exists(marker_dir)) ) { dir.create(marker_dir,recursive = TRUE) }

#############
# LODA DATA #
#############
seuratObj <- readRDS(paste0(input_dir,"seuratObj_clustered.RDS"))

```

```{r Functions}
#######################
# PLOT GENES FUNCTION #
#######################
# obj <- seuratObj
 # gene <- sym("CTSK")
plot_genes.fun <- function(obj, gene, mins=NULL, maxs=NULL, red = "umap_harmony", lable = TRUE){
  gene <- sym(gene)
  obj <- obj %>%
    mutate(lab = obj@active.ident) %>%
    mutate(., FetchData(., vars = c(as_label(gene))) ) %>%
    mutate(feat = !!(gene))
  feat_vec <- pull(obj, as_label(gene))
  
  # Colour pal:
  if(is.null(mins)){
    mins <- min(c(feat_vec, 0),na.rm = T)} # get 0 or negative value
  if(is.null(maxs)){maxs <- quantile(feat_vec,0.99,na.rm = T) # get percentile
    if(maxs==0){maxs <- max(feat_vec,na.rm = T)}
  }
  if(max(feat_vec, na.rm=T) != 0){
    # Calculate percentage:
    obj <- obj %>%
      mutate(feat = (!!(gene) - mins) / ( maxs - mins) ) %>%
      mutate(feat = ifelse(.$feat > 1, 1, .$feat))
  }
  
  obj <- obj %>%
      #select(1:3, !!(gene)) %>%
      mutate(feat = round(.$feat*98)+1) %>%
      mutate(pal = c( col[1],colorRampPalette(col[-1])(99))[.$feat] ) %>%
      arrange(!!(gene))
  
  # reduction method:
  red_1 <- sym(paste0(red, "_1"))
  red_2 <- sym(paste0(red, "_2"))
  
  # txt lable:
  if(lable == FALSE){l = FALSE
    text <- NoLegend() #+ labs(color= "Clusters")
  }else{l = TRUE
    if(lable != TRUE){obj <- mutate(obj, lab = pull(obj, lable))}
    
    lable_df <- obj %>%
      group_by(lab) %>%
      select(lab, contains(red)) %>% 
      summarize_all(mean) 
    
    text <- geom_text(data = lable_df, aes(label = lab), col="black", size=2.5) }
  
  p <- ggplot(obj, aes(!!(red_1), !!(red_2), label=l , color = pal) ) +
    geom_point(alpha = 0.5, size=.5) + ggtitle(as_label(gene)) +
    text + #scale_color_viridis(option = "D", na.value="#EBECF0") +
    scale_colour_identity() +
    my_theme + theme_void() +
    theme(legend.position = "bottom",
          plot.title = element_text(hjust = 0.5)) 
  return(p)
}

##########################
# PLOT CLUSTERS FUNCTION #
##########################
# obj <- seuratObj
# cluster <- sym("RNA_snn_res.0.1")
plot_clusters.fun <- function(obj, cluster, red = "umap_harmony", color = "Brew_all", lable = TRUE){
  if(color == "Brew_all"){
    pal <- c(scales::hue_pal()(8),
             RColorBrewer::brewer.pal(9,"Set1"),
             RColorBrewer::brewer.pal(8,"Set2"),
             RColorBrewer::brewer.pal(8,"Accent"),
             RColorBrewer::brewer.pal(9,"Pastel1"),
             RColorBrewer::brewer.pal(8,"Pastel2") )}
  
  cluster <- sym(cluster)
  
  if(lable == TRUE){ lab <- cluster
    text <- NoLegend() #+ labs(color= "Clusters")
  }else{lab <- sym(lable)
    text <- guides(color = "none")}
  
  feat <- obj %>%
    select(.cell, !!(cluster), !!(lab), sample_name, nCount_RNA, nFeature_RNA) %>%
    group_by(!!(cluster)) %>%
    add_tally() %>%
    arrange(nFeature_RNA) %>%
    arrange(desc(n))
  
  lable_df <- feat %>%
    ungroup() %>%
    group_by(!!(lab)) %>%
    select(!!(lab), contains(red)) %>% 
    summarize_all(mean)
  
  red_1 <- sym(paste0(red, "_1"))
  red_2 <- sym(paste0(red, "_2"))
  
  p <- ggplot(feat, aes(!!(red_1), !!(red_2), 
                        color = !!cluster), label=TRUE) + 
    geom_point(alpha = 0.5, size=.5) + ggtitle(as_label(cluster)) +
    geom_text(data = lable_df, aes(label = !!(lab)), col="black", size=2.5) +
    #guides(color = guide_legend(override.aes = list(size=2, alpha = 1))) +
    scale_color_manual(values = pal)  +
    my_theme + text
  return(p)
}
```

```{r ggplot_theme}
my_theme <-
  list(
    #scale_fill_manual(values = friendly_cols),
    #scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 12),
        legend.position = "bottom",
        #aspect.ratio = 1,
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
      )
  )
```

## Identify marker genes to seperate clusters
```{r Find_marker_genes}
# Compute differential expression one clusters against rest 
markers_genes <- seuratObj %>%
  FindAllMarkers(.,
    log2FC.threshold = 0.1, 
    test.use = "wilcox",
    min.pct = 0.1, 
    min.diff.pct = 0.1, 
    only.pos = TRUE, 
    max.cells.per.ident = 50,
    assay = "RNA", slot = "data")


#markers_genes$pct.diff <- -markers_genes$pct.2-markers_genes$pct.1
#markers_genes$log.pct.diff <- -log2(markers_genes$pct.2/markers_genes$pct.1)
```

## Save seurat object
```{r save_SeuratObject}
###########################
# SAVE INTERMEDIATE OJECT #
###########################
# saveRDS(markers_genes, paste0(result_dir,"markers_genes.RDS"))
# markers_genes <- readRDS(paste0(result_dir,"markers_genes.RDS"))
```

```{r Get_best_markrs}
# Get difference in expression between one cluster and the rest
markers_genes <- markers_genes %>%
  mutate(pct.diff = -markers_genes$pct.2-markers_genes$pct.1) %>%
  mutate(log.pct.diff = -log2(markers_genes$pct.2/markers_genes$pct.1))

# Identify the top genes that have a high difference in expression
top25 <- markers_genes %>%
  group_by(cluster) %>%
  top_n(-40, p_val_adj) %>%
  top_n(20, pct.diff) %>%
  top_n(10, log.pct.diff) 

# remove all VDJ-genes from list of HVG
remove <- str_subset(top25$gene, "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG")
top25_gr <- top25 %>%
  ungroup() %>%
  filter(., !(.$gene %in% remove)) %>%
  group_by(cluster) %>%
  group_split() %>%
  set_names(., seq(0, length(.)-1) )
  

top25 <- top25 %>%
  arrange(cluster)
# top25[grep("NCR1",top25$gene),]
```


```{r 04a_barplot_top25, fig.width=10, , fig.height=6}
pdf(paste0(marker_dir,"top_DEG.pdf"), #width = 4*10*300, height = 30*4*300#, res = 300
    )
par(mfrow=c(2, 5), mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
dev.off()

par(mfrow=c(2, 5), mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)[1:10]) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
```

```{r graph_plot_top25, fig.width=10, fig.height=10}
col=c("grey90","grey80","grey60","navy","black")

clus_plot <- plot_clusters.fun(seuratObj, cluster="RNA_snn_res.1.5") + theme_void() + NoLegend()

grid_genes <- function(plot_list, title){
  title <- ggdraw() + draw_label(paste0("Top 10 Markers for Cluster ", title), fontface='bold')
  g <- plot_grid(plotlist = plot_list,
            ncol = 4)
  g_ <- plot_grid(title, g, ncol=1, rel_heights=c(0.1, 1))
  return(g_)
}

plots <- imap(top25_gr, "gene") %>%
  map(., map, ~plot_genes.fun(seuratObj, .x, lable = "RNA_snn_res.1.5"))

cluster_markers <- plots %>%
  map(., ~c("Clusters"=list(clus_plot), .x)) %>%
  imap(., ~grid_genes(.x, .y ))

cluster_markers[[1]]

imap(cluster_markers, ~ggsave(paste0(marker_dir,"Marker_genes_cluster_", .y, ".jpg"), plot=.x))
# ggsave("multipage.pdf", gridExtra::marrangeGrob(grobs=cluster_markers, ncol=1, nrow=1))
```

# Paulos Code
```{r eval=FALSE, include=FALSE}
library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)

for(df in top25_gr){
  for(gene in df){
    png(paste0(name,"out_gr.png"), width = 4*10*300, height = 30*4*300, res = 300)
  #par(mfrow=c(30,10), mar=c(1,1,1,1))
    par(mfrow=c(6,4), mar=c(1,1,1,1))
    plot_feat(seuratObj, 
              red = "umap_harmony", 
              feat = gene, cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
  }
}
dev.off()

plot_feat(seuratObj, 
              red = "umap_harmony", 
              feat = "CTSK", cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
```

```{r eval=FALSE, include=FALSE}
g <- c("CCL2", "COL1A1", "SYTL3", "IGFBP4", "PTGDS", "LTB")
genes <- c("CD8A", "MYOZ2", "CD3E", "EPCAM", "COL6A1", "CD4", "JCHAIN", "KRT13")
Fibroblast <- c("APOD", "DCN")
NK <- c()

library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)
png("out_.png", width = 4*10*300, height = 30*4*300, res = 300)
par(mfrow=c(30,10), mar=c(1,1,1,1))
#par(mfrow=c(2,4), mar=c(1,1,1,1))
for(i in top25$gene[1:200]){
  plot_feat(seuratObj, red = "umap_harmony", feat = i, cex = 0.5, label = "RNA_snn_res.1.5"#, add_legend=T
            )
}
dev.off()

annot <- c(FIB_DCN = "DCN",
           NK_cell = "NCR1",
           )

markers_genes[grep("NCR1",rownames(markers_genes)),]
```

## Session info
```{r}
sessionInfo()
```