---
title: "Differential gene expression"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    code-fold: show
params:
  fig.path: "./Figures/04/"
editor_options: 
  chunk_output_type: console
---

```{r background-job, eval=FALSE, include=FALSE}
source("../../bin/render_with_jobs.R")

# quarto
# render_html_with_job(out_dir = lab_dir)
# fs::file_move(path = file, new_path = paste0(lab_dir, file))

# currently using quarto for github and kniter for html due to source code option 
render_git_with_job(fig_path = "./Figures/04/")
# Change the figure path from ./Figures/03/ to ../Figures/03/:
system2(command = "sed", stdout = TRUE,
        args = c("-i", "''","-e", 's/src=\\"\\./src=\\"\\.\\./g',
                 paste0("./md_files/", basename("./04_DGE_ref_data.md"))))

# kniter
knit_html_with_job(out_dir = "../../lab_book/reference_data/04_DGE_reference_data", fig_path = "./Figures/04/")
```

```{r setup, include=FALSE}
knitr::opts_chunk$set(
    fig.width   = 6.6929133858,
    fig.path    = params$fig.path,
    dev         = c("png"),
    dpi         = 300,
    fig.align   = "center",
    message     = FALSE,
    warning     = FALSE,
    fig.process = function(filename){
      new_filename <- stringr::str_remove(string = filename, 
                                        pattern = "-1")
      fs::file_move(path = filename, new_path = new_filename)
      ifelse(fs::file_exists(new_filename), new_filename, filename)}) 
# setwd("/Users/vilkal/work/Brolidens_work/Projects/Spatial_DMPA/src/00_reference_data")
```

### Load data and libraries
```{r Load-data}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(cowplot)
library(openxlsx)

source("../../bin/plotting_functions.R")

#########
# PATHS #
#########
input_dir <- "../../results/03_clustering_ref_data/"
result_dir <- "../../results/04_DGE_ref_data/"
marker_dir <- "./marker_genes/"
if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
if( isFALSE(dir.exists(marker_dir)) ) { dir.create(marker_dir,recursive = TRUE) }

#############
# LODA DATA #
#############
DATA <- readRDS(paste0(input_dir,"seuratObj_clustered.RDS"))

```

### Identify marker genes to seperate clusters
```{r Find_marker_genes, include=TRUE, eval=FALSE}
# Compute differential expression one clusters against rest 
markers_genes <- DATA %>%
  FindAllMarkers(.,
    log2FC.threshold = 0.1, 
    test.use = "wilcox",
    min.pct = 0.1, 
    min.diff.pct = 0.1, 
    only.pos = TRUE, 
    max.cells.per.ident = 50,
    assay = "RNA", slot = "data")

# Get difference in expression between one cluster and the rest
markers_genes <- markers_genes %>%
  mutate(pct.diff = -markers_genes$pct.2-markers_genes$pct.1) %>%
  mutate(log.pct.diff = -log2(markers_genes$pct.2/markers_genes$pct.1))
```

### Save Marker genes object
```{r save_SeuratObject, include=TRUE, eval=FALSE}
###########################
# SAVE MARKER GENES TABLE #
###########################
saveRDS(markers_genes, paste0(result_dir,"markers_genes.RDS"))
```

### Load allready saved data
```{r get_DEGs, include=TRUE, eval=TRUE}
# Load data
markers_genes <- readRDS(paste0(result_dir,"markers_genes.RDS"))
```

```{r Get-best-markers}
#######################################
# FILTER BY P-VAL logFC AND pct.diff #
######################################
# Identify the top genes that have a high difference in expression
top25 <- markers_genes %>%
  group_by(cluster) %>%
  top_n(-40, p_val_adj) %>%
  top_n(20, pct.diff) #%>%
  #top_n(10, log.pct.diff)

# remove all VDJ-genes from list of HVG
remove <- str_subset(top25$gene, "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG")
top25_gr <- top25 %>%
  top_n(10, log.pct.diff) %>%
  ungroup() %>%
  filter(., !(.$gene %in% remove)) %>%
  group_by(cluster) %>%
  group_split() %>%
  set_names(., seq(0, length(.)-1) )
  

top25 <- top25 %>%
  arrange(cluster)
# top25[grep("NCR1",top25$gene),]
```

```{r 04a_doptplot_marker_genes, fig.width=10, fig.height=length(unique(top25$gene))/6+2}
# dev.new(width=10, height=length(unique(top25$gene))/6+2, noRStudioGD = TRUE) 
#######################################
# DOTPLOT 20 MARKER GENES PER CLUSTER #
#######################################
p <- DotPlot(DATA, features = rev(as.character(unique(top25$gene))), 
        group.by = "Clusters",
    assay = "RNA") + coord_flip()

ggsave(paste0(result_dir, "dotplot.pdf"), p, height = 90, width = 10, limitsize = FALSE)
p
```

```{r 04b_barplot_top25, fig.width=10, fig.height=6*3, dev = c("png", "pdf")}
# dev.new(width=410, height=6, noRStudioGD = TRUE)
###################################
# BARPLOT TOP 20 UPREGULATED DEGs #
###################################
# pdf(paste0(marker_dir,"top_DEG.pdf")) #, width = 4*10*300, height = 30*4*300#, res = 300)
par(mfrow=c(6, 5), mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
# dev.off()
```

```{r 04c_UMAP_top10_marker_genes, fig.width=3*4, fig.height=3*3}
# dev.new(width=3*4, height=3*3, noRStudioGD = TRUE) 
################################
# UMAP TOP 10 DOWN AND UP DEGs #
################################
# col <- c("grey90","grey80","grey60","navy","black")
col <- c("#EFEDF5", "#DADAEB", "#BCBDDC", "#9E9AC8", "#807DBA", "#6A51A3", "#54278F", "#3F007D")

clus_plot <- plot_clusters.fun(DATA, cluster="Clusters") + theme_void() + NoLegend()

grid_genes <- function(plot_list, title){
  title <- ggdraw() + draw_label(paste0("Top 10 Markers for Cluster ", title), fontface='bold')
  g <- plot_grid(plotlist = plot_list,
            ncol = 4)
  g_ <- plot_grid(title, g, ncol=1, rel_heights=c(0.1, 1))
  return(g_)
}

plots <- imap(top25_gr, "gene") %>%
  map(., map, ~plot_genes.fun(DATA, .x, lable = "Clusters", col = col))

cluster_markers <- plots %>%
  map(., ~c("Clusters"=list(clus_plot), .x)) %>%
  imap(., ~grid_genes(.x, .y ))

cluster_markers[[1]]

imap(cluster_markers, ~ggsave(paste0(marker_dir,"Marker_genes_cluster_", .y, ".jpg"), plot=.x, width = 3*4, height = 3*3))
```

### Run this part when you have got a rough speration of the cell types
```{r 04d_doptplot_marker_genes, fig.width=10, fig.height=length(unique(top25$gene))/6+2, eval=TRUE}
# dev.new(width=10, height=length(unique(top25$gene))/6+2, noRStudioGD = TRUE) 
################
# ADD CELLTYPE #
################
cell_id <- read.xlsx("../../../Spatial_DMPA/data/Reference_data/annotation_single_cell.xlsx", sheet = "edited")
cell_type <- set_names(cell_id$annotation_level1, as.character(cell_id$cluster))

#######################################
# DOTPLOT 20 MARKER GENES PER CLUSTER #
#######################################
p <- DATA %>%
  mutate(., cell_type = cell_type[as.character(.$Clusters)]) %>%
  mutate(., annot = paste0(.$cell_type, "_", .$Clusters)) %>%
  mutate(., annot = factor(.$annot)) %>% 
  nest(data = -c("cell_type")) %>%
  mutate(dotplots = map(data, 
                        ~DotPlot(.x, features = rev(as.character(unique(top25$gene))), 
                          group.by = "Clusters", assay = "RNA") + coord_flip()
  ))

pmap(p, ~ggsave(paste0("./Figures/04/", "dotplot_",..1,".pdf"), ..3, height = 90, width = 10, limitsize = FALSE))
```

## Session info
```{r}
sessionInfo()
```

# Paulos Code
```{r eval=FALSE, include=FALSE}
library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)

for(df in top25_gr){
  for(gene in df){
    png(paste0(name,"out_gr.png"), width = 4*10*300, height = 30*4*300, res = 300)
  #par(mfrow=c(30,10), mar=c(1,1,1,1))
    par(mfrow=c(6,4), mar=c(1,1,1,1))
    plot_feat(DATA, 
              red = "umap_harmony", 
              feat = gene, cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
  }
}
dev.off()

plot_feat(DATA, 
              red = "umap_harmony", 
              feat = "CTSK", cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
```

```{r eval=FALSE, include=FALSE}
g <- c("CCL2", "COL1A1", "SYTL3", "IGFBP4", "PTGDS", "LTB")
genes <- c("CD8A", "MYOZ2", "CD3E", "EPCAM", "COL6A1", "CD4", "JCHAIN", "KRT13")
Fibroblast <- c("APOD", "DCN")
NK <- c()

library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)
png("out_.png", width = 4*10*300, height = 30*4*300, res = 300)
par(mfrow=c(30,10), mar=c(1,1,1,1))
#par(mfrow=c(2,4), mar=c(1,1,1,1))
for(i in top25$gene[1:200]){
  plot_feat(DATA, red = "umap_harmony", feat = i, cex = 0.5, label = "Clusters"#, add_legend=T
            )
}
dev.off()

annot <- c(FIB_DCN = "DCN",
           NK_cell = "NCR1",
           )

markers_genes[grep("NCR1",rownames(markers_genes)),]
```
