---
title: "Differential gene expression"
date: "`r format(Sys.time(), '%d %B, %Y')`"
format:
  html:
    embed-resources: true
    code-fold: show
params:
  fig.path: "`r paste0(params$fig.path)`" #./Figures/
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(#fig.width = 6.6929133858,
  fig.path      = params$fig.path,#"../Figures/",
  fig.align     = "center",
  message       = FALSE,
  warning       = FALSE,
  dev           = c("jpeg", "tiff", "pdf"),
  dpi           = 300,
  fig.process = function(filename){
    new_filename <- stringr::str_remove(string = filename,
                                        pattern = "-1")
    fs::file_move(path = filename, new_path = new_filename)
    ifelse(fs::file_exists(new_filename), new_filename, filename)
  }
  )
# setwd("/Users/vilkal/work/Brolidens_work/Projects/Spatial_DMPA/src/00_reference_data")
```

```{r background_job, eval=FALSE}
file <- paste0(basename(xfun::sans_ext("./05_subset_celltypes_ref_data.md")), '_', Sys.Date(), '.html')

# quarto
render_html_with_job(out_dir = "../../lab_book/test/")
fs::file_move(path = file, new_path = paste0("../../lab_book/test/", file))

render_git_with_job()

# kniter
knit_html_with_job(out_dir = "../../lab_book/test/")
```

# Subset celltypes

## Load data and libraries
```{r Load_data}
##################
# LOAD LIBRARIES #
##################
library(tidyverse)
library(Seurat)
library(SeuratObject)
library(tidyseurat)
library(harmony)
library(cowplot)
library(openxlsx)

#########
# PATHS #
#########
input_dir <- "../../results/03_clustering_ref_data/"
result_dir <- "../../results/05_subset_ref_data/"
marker_dir <- "./marker_genes/"
if( isFALSE(dir.exists(result_dir)) ) { dir.create(result_dir,recursive = TRUE) }
if( isFALSE(dir.exists(marker_dir)) ) { dir.create(marker_dir,recursive = TRUE) }

#############
# LODA DATA #
#############
seuratObj <- readRDS(paste0(input_dir,"seuratObj_clustered.RDS"))
cell_id <- read.xlsx("../../data/Reference_data/annotation_single_cell.xlsx")

```

## Subset SeuratObj
```{r subset_cells}
seuratObj_nested <- seuratObj %>%
  left_join(cell_id, by=c("Clusters"="cluster")) %>%
  select(.cell, sample_name, Clusters, annotation_level1:Genes_level3, nCount_RNA:percent_ribo) %>%
  filter(., grepl("fibroblasts|endothelial|epithelial|immune", .$annotation_level1)) %>%
  filter(., ifelse(!(annotation_level1=="immune" & annotation_level2 != "T cell"), TRUE, FALSE)) %>%
  nest(data = -annotation_level1)
```

```{r Functions}
#######################
# PLOT GENES FUNCTION #
#######################
# obj <- seuratObj
 # gene <- sym("CTSK")
plot_genes.fun <- function(obj, gene, mins=NULL, maxs=NULL, red = "umap_harmony", lable = TRUE){
  gene <- sym(gene)
  obj <- obj %>%
    mutate(lab = obj@active.ident) %>%
    mutate(., FetchData(., vars = c(as_label(gene))) ) %>%
    mutate(feat = !!(gene))
  feat_vec <- pull(obj, as_label(gene))
  
  # Colour pal:
  if(is.null(mins)){
    mins <- min(c(feat_vec, 0),na.rm = T)} # get 0 or negative value
  if(is.null(maxs)){maxs <- quantile(feat_vec,0.99,na.rm = T) # get percentile
    if(maxs==0){maxs <- max(feat_vec,na.rm = T)}
  }
  if(max(feat_vec, na.rm=T) != 0){
    # Calculate percentage:
    obj <- obj %>%
      mutate(feat = (!!(gene) - mins) / ( maxs - mins) ) %>%
      mutate(feat = ifelse(.$feat > 1, 1, .$feat))
  }
  
  obj <- obj %>%
      #select(1:3, !!(gene)) %>%
      mutate(feat = round(.$feat*98)+1) %>%
      mutate(pal = c( col[1],colorRampPalette(col[-1])(99))[.$feat] ) %>%
      arrange(!!(gene))
  
  # reduction method:
  red_1 <- sym(paste0(red, "_1"))
  red_2 <- sym(paste0(red, "_2"))
  
  # txt lable:
  if(lable == FALSE){l = FALSE
    text <- NoLegend() #+ labs(color= "Clusters")
  }else{l = TRUE
    if(lable != TRUE){obj <- mutate(obj, lab = pull(obj, lable))}
    
    lable_df <- obj %>%
      group_by(lab) %>%
      select(lab, contains(red)) %>% 
      summarize_all(mean) 
    
    text <- geom_text(data = lable_df, aes(label = lab), col="black", size=2.5) }
  
  p <- ggplot(obj, aes(!!(red_1), !!(red_2), label=l , color = pal) ) +
    geom_point(alpha = 0.5, size=.5) + ggtitle(as_label(gene)) +
    text + #scale_color_viridis(option = "D", na.value="#EBECF0") +
    scale_colour_identity() +
    my_theme + theme_void() +
    theme(legend.position = "bottom",
          plot.title = element_text(hjust = 0.5)) 
  return(p)
}

##########################
# PLOT CLUSTERS FUNCTION #
##########################
# obj <- seuratObj
# cluster <- sym("RNA_snn_res.0.1")
plot_clusters.fun <- function(obj, cluster, red = "umap_harmony", color = "Brew_all", lable = TRUE){
  if(color == "Brew_all"){
    pal <- c(scales::hue_pal()(8),
             RColorBrewer::brewer.pal(9,"Set1"),
             RColorBrewer::brewer.pal(8,"Set2"),
             RColorBrewer::brewer.pal(8,"Accent"),
             RColorBrewer::brewer.pal(9,"Pastel1"),
             RColorBrewer::brewer.pal(8,"Pastel2") )}
  
  cluster <- sym(cluster)
  
  if(lable == TRUE){ lab <- cluster
    text <- NULL#NoLegend() #+ labs(color= "Clusters")
  }else{lab <- enquo(lable)
    text <- guides(color = "none")}
  
  feat <- obj %>%
    select(.cell, !!(cluster), !!(lab), sample_name, nCount_RNA, nFeature_RNA) %>%
    group_by(!!(cluster)) %>%
    add_tally() %>%
    arrange(nFeature_RNA) %>%
    arrange(desc(n))
  
  lable_df <- feat %>%
    ungroup() %>%
    group_by(!!(lab)) %>%
    select(!!(lab), contains(red)) %>% 
    summarize_all(mean)
  
  red_1 <- sym(paste0(red, "_1"))
  red_2 <- sym(paste0(red, "_2"))
  
  p <- ggplot(feat, aes(!!(red_1), !!(red_2), 
                        color = !!cluster), label=TRUE) + 
    geom_point(alpha = 0.5, size=.5) + ggtitle(as_label(cluster)) +
    geom_text(data = lable_df, aes(label = !!(lab)), col="black", size=2.5) +
    guides(color = guide_legend(override.aes = list(size=2, alpha = 1))) +
    scale_color_manual(values = pal)  +
    my_theme + text
  return(p)
}

################
# GGPLOT THEME #
################
my_theme <-
  list(
    #scale_fill_manual(values = friendly_cols),
    #scale_color_manual(values = friendly_cols),
    theme_bw() +
      theme(
        panel.border = element_blank(),
        axis.line = element_line(),
        panel.grid.major = element_line(size = 0.2),
        panel.grid.minor = element_line(size = 0.1),
        text = element_text(size = 12),
        legend.position = "bottom",
        #aspect.ratio = 1,
        strip.background = element_blank(),
        axis.title.x = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10)),
        axis.title.y = element_text(margin = margin(t = 10, r = 10, b = 10, l = 10))
      )
  )
```

```{r Find_HVG, eval=FALSE, include=FALSE}
################################
# SPLIT INTO SEPERATE DATASETS #
################################
seuratObj_nested <- seuratObj %>%
  mutate(subset = annotation_level1) %>% 
  nest(data = -annotation_level1)%>%
  mutate(data = imap(
    data, ~ .x %>%
      #NormalizeData(., normalization.method = "LogNormalize",verbose = FALSE) %>%
      FindVariableFeatures(selection.method = "vst", 
                           nfeatures = 2000, 
                           verbose = FALSE) )) %>%
  mutate(data = setNames(.[["data"]], .$annotation_level1))

#########################################
# FIND HIGLY VARIABLE GENES PER SUBSET #
#########################################
hig_var <- seuratObj_nested %>%
  .$data %>%
  map(., ~ unique(unlist(.x@assays$RNA@var.features))) %>%
  # remove all VDJ-genes from list of HVG:
  map(., ~ setdiff(.x, str_subset(.x, "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG")))


hig_var.fun <- function(Obj){
  hig_var <- Obj@assays$RNA@var.features %>%
    # remove all VDJ-genes from list of HVG:
    setdiff(., str_subset(., "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG"))
  Obj@assays$RNA@var.features <- hig_var
  return(Obj)
}


###########################
# DIMENTIONAITY REDUCTION #
###########################
seuratObj_nested <- seuratObj_nested %>%
  .$data %>%
  map2(., hig_var, ~ ScaleData(.x, verbose = FALSE, features = .y )) %>%
  map2(.,  ~ RunPCA(verbose = FALSE, npcs = 50)) %>%
  map2(.,  ~ RunUMAP(dims = 1:50, n.components = 2L) )

# seuratObj <- seuratObj_nested$data$fibroblasts %>%
#   ScaleData(verbose = FALSE, features = hig_var$fibroblasts ) %>%
#   RunPCA(verbose = FALSE, npcs = 50) %>%
#   RunUMAP(dims = 1:50, n.components = 2L) 
```

## Identify Highly Variable Genes (HVG) in each subset
```{r HVG}
#########################################
# FIND HIGLY VARIABLE GENES PER SUBSET #
#########################################
hig_var.fun <- function(Obj){
  hig_var <- Obj@assays$RNA@var.features %>%
    # remove all VDJ-genes from list of HVG:
    setdiff(., str_subset(., "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG"))
  Obj@assays$RNA@var.features <- hig_var
  return(Obj)
}

###########################
# DIMENTIONAITY REDUCTION #
###########################
seuratObj_nested <- seuratObj_nested[1:2,] %>%
  mutate(data = map(
    data, ~ .x %>%
      FindVariableFeatures(verbose = FALSE) %>%
      hig_var.fun(.) %>%
      RunPCA(npcs = 20, verbose = FALSE) %>%
      RunHarmony(group.by.vars = "sample_name", 
             reduction = "pca", 
             dims.use = 1:20, 
             assay.use = "RNA") %>%
      RunUMAP(reduction = "harmony", dims = 1:20, 
              n.components = 2L, verbose = FALSE) %>%
      FindNeighbors(verbose = FALSE, reduction = "harmony", dims = 1:20) )) #%>%
      #FindClusters( verbose = FALSE, resolution = 0.8)

# seuratObj_fibro <- seuratObj_nested$data$fibroblasts %>%
#   FindVariableFeatures(verbose = FALSE) %>%
#   hig_var.fun(.) %>%
#   RunPCA(npcs = 10, verbose = FALSE) %>%
#   FindNeighbors(verbose = FALSE) %>%
#   FindClusters(method = "igraph", verbose = FALSE) %>%
#   RunUMAP(reduction = "pca", dims = 1:10, n.components = 3L, verbose = FALSE)

```

## Save seurat object
```{r save_SeuratObject_}
###########################
# SAVE INTERMEDIATE OJECT #
###########################
saveRDS(seuratObj_nested, paste0(result_dir,"subset_celltypes.RDS"))
# seuratObj_nested <- readRDS(paste0(result_dir,"subset_celltypes.RDS"))
```

## Clustering
```{r Clustering_resolution}
##################################
# EVALUATE CLUSTERING RESOLUTION #
##################################
# Clustering with louvain (algorithm 1) or leiden (algorithm 4)
clus_res.fun <- function(obj){
  for (res in c(0.1, 0.25, 0.5, 1, 1.5, 2)) {
    obj <- FindClusters(obj, resolution = res, algorithm = 1)
  }
  return(obj)
}

seuratObj_nested <- seuratObj_nested %>%
  mutate(data = map(
    data, ~ .x %>%
      clus_res.fun(.) ))

# each time you run clustering, the data is stored in meta data columns:
# seurat_clusters - lastest results only CCA_snn_res.XX - for each different
# resolution you test.

```

```{r, fig.width=10}
#res <- c("RNA_snn_res.0.1", "RNA_snn_res.0.25")
res <- c("RNA_snn_res.0.1", "RNA_snn_res.0.25",
         "RNA_snn_res.0.5","RNA_snn_res.1", "RNA_snn_res.1.5")
clus_plots <- seuratObj_nested %>%
  add_column(., Res = list(res)) %>%
  rowwise() %>%
  mutate(Plots = list(map(Res, ~plot_clusters.fun(data, cluster=.x, red = "UMAP"))) )

clus_plots <- clus_plots %>%
  rowwise() %>%
  mutate(., plot_grid = list(plot_grid(plotlist = Plots)))

clus_plots$plot_grid[[1]]
clus_plots$plot_grid[[2]]
#clus_plots$plot_grid[[3]]
#clus_plots$plot_grid[[4]]

res <- c(fibroblasts="RNA_snn_res.0.1", immune="RNA_snn_res.0.25"#,endothelial="RNA_snn_res.0.25", epithelial="RNA_snn_res.0.5"
         )

seuratObj_nested <- seuratObj_nested %>%
  mutate(name = res) %>%
  mutate(data = imap(
    data, ~ .x %>%
      select(.cell, sample_name, Clusters_old="Clusters", 
             Clusters=name[[.y]], everything()) %>%
      SetIdent(., value = "Clusters") %>%
      select(-starts_with("RNA_snn_res."))))

seuratObj_nested_$data[[2]]
```

```{r tidySeurat_plot}
seuratObj_nested %>%

  # Convert to tibble otherwise Seurat drops reduced dimensions when unifying data sets.
  mutate(data = map(data, ~ .x %>% as_tibble())) %>%
  unnest(data) %>%

  # Define unique clusters
  unite("cluster", c(annotation_level1, Clusters), remove = FALSE) %>%

  # Plotting
  ggplot(aes(UMAP_1, UMAP_2, color = cluster)) +
  geom_point() +
  facet_wrap(~annotation_level1) +
  my_theme
```

```{r, fig.width=10}
clus_plots <- seuratObj_nested %>%
  mutate(data = map(
    data, ~ .x %>%
      mutate(Clusters = factor(.$Clusters)) %>%
      plot_clusters.fun(., cluster="Clusters", red = "UMAP")) ) # "umap_harmony"

fibro <- clus_plots$data[[1]]# + theme_void() + NoLegend()
T_cell <- clus_plots$data[[2]]
endo <- clus_plots$data[[3]]
epi <- clus_plots$data[[4]]
clus <- plot_clusters.fun(seuratObj, cluster="Clusters", red = "umap_harmony")

plot_grid(clus, fibro)
```

## Identify marker genes to seperate clusters
```{r Find_marker_genes}
# Compute differential expression one clusters against rest 
markers_genes <- seuratObj_nested %>%
  mutate(data = map(
    data, ~ .x %>%
    FindAllMarkers(.,
      log2FC.threshold = 0.1, 
      test.use = "wilcox",
      min.pct = 0.1, 
      min.diff.pct = 0.1, 
      only.pos = TRUE, 
      max.cells.per.ident = 50,
      assay = "RNA", slot = "data") ) %>%
      set_names(., seuratObj_nested$annotation_level1))

#markers_genes$pct.diff <- -markers_genes$pct.2-markers_genes$pct.1
#markers_genes$log.pct.diff <- -log2(markers_genes$pct.2/markers_genes$pct.1)
```

## Save marker genes object
```{r save_marker_genes}
###########################
# SAVE INTERMEDIATE OJECT #
###########################
# saveRDS(markers_genes, paste0(result_dir,"markers_genes.RDS"))
# markers_genes <- readRDS(paste0(result_dir,"markers_genes.RDS"))
```

```{r Get_best_markers}
# Get difference in expression between one cluster and the rest
markers_genes <- markers_genes %>%
  mutate(data = map(
    data, ~ .x %>%
      mutate(pct.diff = -.$pct.2-.$pct.1) %>%
      mutate(log.pct.diff = -log2(.$pct.2/.$pct.1)) ))

# Identify the top genes that have a high difference in expression
remove <- "^IGH|^IGK|^IGL|^TRA|^TRB|^TRD|^TRG"
markers_genes <- markers_genes %>%
  mutate(top25 = map(
    data, ~ .x %>%
      group_by(cluster) %>%
      top_n(-40, p_val_adj) %>%
      top_n(30, pct.diff) %>%
      top_n(20, log.pct.diff) %>%
      
      ungroup() %>%
      # remove all VDJ-genes from list of HVG:
      filter(., !(.$gene %in% str_subset(.$gene, remove)))
    ))
    
# top25 <- markers_genes$data$fibroblasts %>%
#   group_by(cluster) %>%
#   top_n(-40, p_val_adj) %>%
#   top_n(30, pct.diff) %>%
#   top_n(20, log.pct.diff) 

markers_genes <- markers_genes %>%
  mutate(top25_gr = map(
    top25, ~ .x %>%
    group_by(cluster) %>%
    group_split() %>%
    set_names(., seq(0, length(.)-1) ) ))
  
# top25[grep("NCR1",top25$gene),]
markers_genes
```


```{r Save_dotplot}
top25 <- markers_genes$top25$fibroblasts

gene_list <- map2(seuratObj_nested$data,markers_genes$top25, 
                  ~niceRplots::getcluster(.x,as.character(unique(.y$gene)),
                                          "Clusters"))
immune <- seuratObj_nested$data[[2]]
genes_immune <- rownames(immune)[grepl( "CD3|TR[ABGD]C|IL[0-9]|CC[LR][0-9]",rownames(immune))] %>%
genes_immune <- getcluster(data = immune, clustering =  "Clusters", genes = gene_list)

library(niceRplots)
gene_list2 <- 

p <- map2(list(seuratObj_nested$data[[2]]), list(gene_list2),
          ~DotPlot(.x, features = rev(names(.y)[order(.y)]),
                   group.by = "Clusters",
                   assay = "RNA") + coord_flip() ) #%>%
  #set_names(., seuratObj_nested$annotation_level1)

walk(p, ~ggsave(paste0("./Figures/05_subset_celltypes/","dotplot_", "immune_related_genes",".pdf"), .x, height = 30, width = 7, limitsize = FALSE))
```


```{r 04a_barplot_top25, fig.width=10, fig.height=6}
pdf(paste0(marker_dir,"top_DEG.pdf"), #width = 4*10*300, height = 30*4*300#, res = 300
    )
par(mfrow=c(2, 5), mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
dev.off()

par(mfrow=c(2, 5), mar = c(4, 6, 3, 1))
for (i in unique(top25$cluster)[1:10]) {
    barplot(sort(setNames(top25$avg_log2FC, top25$gene)[top25$cluster == i], F),
        horiz = T, las = 1, main = paste0(i, " vs. rest"), border = "white", yaxs = "i")
    abline(v = c(0, 0.25), lty = c(1, 2))
}
```

```{r graph_plot_top25, fig.width=10, fig.height=10}
col=c("grey90","grey80","grey60","navy","black")

clus_plot <- plot_clusters.fun(seuratObj, cluster="RNA_snn_res.1.5") + theme_void() + NoLegend()

grid_genes <- function(plot_list, title){
  title <- ggdraw() + draw_label(paste0("Top 10 Markers for Cluster ", title), fontface='bold')
  g <- plot_grid(plotlist = plot_list,
            ncol = 4)
  g_ <- plot_grid(title, g, ncol=1, rel_heights=c(0.1, 1))
  return(g_)
}

plots <- imap(top25_gr, "gene") %>%
  map(., map, ~plot_genes.fun(seuratObj, .x, lable = "RNA_snn_res.1.5"))

cluster_markers <- plots %>%
  map(., ~c("Clusters"=list(clus_plot), .x)) %>%
  imap(., ~grid_genes(.x, .y ))

cluster_markers[[1]]

imap(cluster_markers, ~ggsave(paste0(marker_dir,"Marker_genes_cluster_", .y, ".jpg"), plot=.x))
# ggsave("multipage.pdf", gridExtra::marrangeGrob(grobs=cluster_markers, ncol=1, nrow=1))
```

# Paulos Code
```{r eval=FALSE, include=FALSE}
library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)

for(df in top25_gr){
  for(gene in df){
    png(paste0(name,"out_gr.png"), width = 4*10*300, height = 30*4*300, res = 300)
  #par(mfrow=c(30,10), mar=c(1,1,1,1))
    par(mfrow=c(6,4), mar=c(1,1,1,1))
    plot_feat(seuratObj, 
              red = "umap_harmony", 
              feat = gene, cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
  }
}
dev.off()

plot_feat(seuratObj, 
              red = "umap_harmony", 
              feat = "CTSK", cex = 0.5, 
              label = "RNA_snn_res.1.5"
              #, add_legend=T
            )
```

```{r eval=FALSE, include=FALSE}
g <- c("CCL2", "COL1A1", "SYTL3", "IGFBP4", "PTGDS", "LTB")
genes <- c("CD8A", "MYOZ2", "CD3E", "EPCAM", "COL6A1", "CD4", "JCHAIN", "KRT13")
Fibroblast <- c("APOD", "DCN")
NK <- c()

library(niceRplots)
#pdf("out.pdf", width = 4*10, height = 27*4, useDingbats = F)
png("out_.png", width = 4*10*300, height = 30*4*300, res = 300)
par(mfrow=c(30,10), mar=c(1,1,1,1))
#par(mfrow=c(2,4), mar=c(1,1,1,1))
for(i in top25$gene[1:200]){
  plot_feat(seuratObj, red = "umap_harmony", feat = i, cex = 0.5, label = "RNA_snn_res.1.5"#, add_legend=T
            )
}
dev.off()

annot <- c(FIB_DCN = "DCN",
           NK_cell = "NCR1",
           )

markers_genes[grep("NCR1",rownames(markers_genes)),]
```

## Session info
```{r}
sessionInfo()
```